docs/GDPs/gdp-0002.rst

```diff
@@ -0,0 +1,346 @@
+
+
+Compilation-Time Evaluation
++++++++++++++++++++++++++++
+
+The expression `expr` inside `region(expr)` is evaluated at compile-time, and the resulting index ranges are stored in the library module created.
+Importantly, this means that GT4Py does not need to create objects to mirror these expression at call-time, as they may call partitioners or other application-specific code.
+
+
+Box Objects
++++++++++++
+
+The `Box` slice data could be coverted into a `local_origin` and a `local_domain`, to mirror the existing implementation and stay compatible with GridTools boundary conditions.
+If the `local_domain` is of size `0` in any dimension, the block will not execute.
+This information could be utilized by the backend to optimize not computing unnecessary stencils.
+Note that `region(*args)` so the same sub-stencil can execute on any number of boxes passed, as was shown in the motivating example.
+
+Application developers will create their own specializations of Box for their particular uses.
+For example, since our use case has edges and corners, we may have specific `Corner` specializations as well.
+
+When using `region()`, the stencil caller should avoid using `origin` and `domain` as this would change the iteration space and would be very error prone.
+
+Halo Regions
+++++++++++++
+
+It is important for our application that halo regions can be accessed by the stencils.
+This falls out naturally from the `region()` by passing negative integers to the `Edge` boxes.
+For example, `WestEdge(-2,0)` will access two lines of halo points.
+
+
+Corner Cases and Warnings
++++++++++++++++++++++++++
+
+It may be the case that a part of the iteration space is accessed twice by different `with` blocks.
+These should ideally be scheduled and run in the order they appear, but as a first implementation GT4Py could throw an error in this case.
+An option should be added to determine if this should be an error.
+
+There could also be an option to warn the user if they have not covered the entire iteration space in a stencil.
+This could be useful for debugging.
+
+
+Related Work
+------------
+
+The dawn compiler accepts regions of the horizontal iteration space in its stencil intermediate representation.
+
+
+Implementation
+--------------
+
+The implementation on the GT4Py side should be relatively light. We will need to
+
+1. add the `region()` specifier parsing to the AST visitor in IRMaker_
+2. add the reduced iteration space to the internal IR
+3. ensure backends parse this information correctly
+4. add a few corresponding tests for the feature
+
+.. _IRMaker: https://github.com/GridTools/gt4py/blob/master/src/gt4py/frontend/gtscript_frontend.py#L454
+
+Implementation on the application side is to setup specializations of the Box class for its needs.
+
+It would be interesting to perform math operations like `intersection()` and `union()` of boxes.
+This can be implemented in a later step purely on the GT4Py side.
```

Comment:
We can't actually perform union on these region objects, since the union of rectangular regions is not necessarily rectangular. Union is implied implicitly by providing multiple arguments to gt4py. Intersection makes more sense to provide as a pure-python function, which could be used in providing arguments to `region`. I think though it's out of scope for this GDP.
```suggestion
```

Reply:
Agreed. We can just suggest this feature eventually as an add-on.